### PR TITLE
Add base elements to support distributed comms. Add supports_distributed plugin flag.

### PR DESCRIPTION
**Pull request by lrzpellegrini**
This PR ports elements from the #996 PR by adding the following:
- elements such DistributedHelper that, without changing the current behavior of Avalanche elements, is the starting point to start wiring the support for distributed training
- the `supports_distributed` flag in BasePlugin. The method used to inherit such a flag in child classes is different from the one in #996 
- A change in the data type used to store masks in multi-task modules

Unrelated to distributed training:
- Adds the `_obtain_common_dataloader_parameters` in strategy
- Minor fixes to the multi-task dynamic modules, that did not consider certain corner cases regarding evaluation before training
- Fixes for the LVISDataset
- Removed a few unused imports / added missing type hints / fixed typos / ...

Part of the effort described here: #1315

### CODE REVIEW

**Review by AntonioCarta — CHANGES_REQUESTED** [review]

**Comment by AntonioCarta** [comment]
Thanks @lrzpellegrini, I added a bunch of comments. I think there are some parts that can be simplified a bit hopefully.

File: avalanche/distributed/distributed_consistency_verification.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+if TYPE_CHECKING:
+    from torch import Tensor
+    from torch.nn import Module
+    from avalanche.benchmarks import DatasetScenario
+    from torch.utils.data import Dataset
+
+
+def hash_benchmark(benchmark: 'DatasetScenario', *, 
+                   hash_engine=None, num_workers=0) -> str:
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/distributed/distributed_consistency_verification.py:18]: Yes, I think I can move those elements to the appropriate classes.

File: avalanche/distributed/distributed_helper.py [hunk 2]
```diff
@@ -0,0 +1,819 @@
...
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(_Singleton, cls).__call__(
+                *args, **kwargs)
+        return cls._instances[cls]
+
+
+class RollingSeedContext(object):
+    """
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/distributed/distributed_helper.py:36]: Those contexts are only to manage the situation in which some processes (like the main one) may consume random numbers in a different way from other processes. It ensures that the random number generators are aligned across processes after exiting a certain section wrapped using `with`:

```python
with seed_context():
    use_rng()
```

File: avalanche/distributed/distributed_helper.py [hunk 3]
```diff
@@ -0,0 +1,819 @@
...
+            # Let other process enter the section
+            DistributedHelper.barrier()
+
+        self._seed_aligner.__exit__(exc_type, exc_val, exc_tb)
+        if self._final_barrier:
+            DistributedHelper.barrier()
+
+
+class _DistributedHelperCls(object):
+    """
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/distributed/distributed_helper.py:122]: The problem with torch.distributed is that it is a terribly low-level API, especially when you want to support both "cpu" and "cuda" devices. Using the helper is not mandatory and users will be able to use the torch.distributed API as always. The methods implemented in the helper are to simplify code readability.

File: avalanche/logging/base_logger.py [hunk 4]
```diff
@@ -28,6 +30,31 @@ class BaseLogger(ABC):
     def __init__(self):
         super().__init__()
 
+        if not DistributedHelper.is_main_process:
+
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/logging/base_logger.py:34]: The problem is that operations done by loggers will end up overlapping in most cases (not only file-system access, but also logging on cloud dashboards such as W&B) with problematic results. Most elements to be monitored per-process, mainly resource usage (as CPU/GPU) usage, should be implemented as a metric. One of the steps for the next PRs it to properly adapt performance metrics to allow the main process to gather and log those values.

File: avalanche/models/dynamic_modules.py [hunk 5]
```diff
@@ -75,6 +75,11 @@ def eval_adaptation(self, experience: CLExperience):
         """
         pass
 
+    @property
+    def model_device(self):
+        """Returns the device of the model."""
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/models/dynamic_modules.py:80]: I agree, but we need a default device to use when creating the masks.

File: avalanche/models/dynamic_modules.py [hunk 6]
```diff
@@ -237,7 +243,10 @@ def adaptation(self, experience: CLExperience):
         if self.masking:
             if old_nclasses != new_nclasses:  # expand active_units mask
                 old_act_units = self.active_units
-                self.active_units = torch.zeros(new_nclasses, dtype=torch.bool)
+                self.active_units = torch.zeros(
+                    new_nclasses,
+                    dtype=torch.int8, 
+                    device=device)
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/models/dynamic_modules.py:249]: torch.bool is not supported by DistributedDataParallel. I too was a little bummed when I found that out. With int8 it doesn't complain.

File: avalanche/training/templates/base.py [hunk 7]
```diff
@@ -39,7 +41,7 @@ class BaseTemplate(BaseStrategyProtocol[TExperienceType]):
     def __init__(
         self,
         model: Module,
-        device: Union[str, torch.device] = "cpu",
+        device: Optional[Union[str, torch.device]] = "cpu",
         plugins: Optional[Sequence[BasePlugin]] = None,
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/training/templates/base.py:45]: Alas, we have a lot of code in the wild passing `None` to the device parameter. However, we can remove the Optional type hint to state that it should not be None from now on (also maybe raising a warning at runtime).

File: avalanche/core.py [hunk 8]
```diff
@@ -68,6 +78,18 @@ def after_eval(self, strategy: Template, *args, **kwargs) -> CallbackResult:
         """Called after `eval` by the `BaseTemplate`."""
         pass
 
+    def _check_distributed_support(
+            self,
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/core.py:82]: That is a bit tricky. The problem with an inherited class-level flag is that child classes of a plugin may not actually support distributed training (differently from the father class).

The idea is that the father class sets the default value of the constructor parameter `supports_distributed` to None. Then, 2 situations may happen:

1. If `self` is an instance of a child class
  - If the child constructor passes True/False as a value to `supports_distributed` to the super constructor, ten that value will be used as-is.
  - If it does not pass any value, it means that `_check_distributed_support` is called with `None` and it will detect that the child class did not explicitly support distributed training, returning False.
2. `self` is an instance of the "current" class: `_check_distributed_support` is called with `None`, detects that the father class supports distributed training and returns True.

In this way, for a plugin class to not flag itself as supporting distributed training, it is sufficient to call the super constructor by not passing the `supports_distributed` parameter at all. This is good because we can mark existing plugins as not explicitly supporting distributed training without the need to change their code.

File: avalanche/training/templates/base_sgd.py [hunk 9]
```diff
@@ -345,11 +345,45 @@ def _before_eval_exp(self, **kwargs):
 
         super()._before_eval_exp(**kwargs)
 
+    def _obtain_common_dataloader_parameters(self, **kwargs):
+        """
+        Utility function that returns the dictionary of parameters to be passed
+        to the train and eval dataloaders.
+
+        The resulting dataset does not include the collate function.
+
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/training/templates/base_sgd.py:354]: Yes, it's a typo. It should be `The resulting dictionary ...`

File: avalanche/training/templates/base_sgd.py [hunk 10]
```diff
@@ -345,11 +345,45 @@ def _before_eval_exp(self, **kwargs):
...
+        Overriding this function can be useful if particular/runtime computed
+        parameters are needed. However, when overriding, it is recommended to
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/training/templates/base_sgd.py:356]: I think it can be rephrased a bit to suggest that, if changes are limited to a few dataloader parameters, it should be enough to override this function, while if the changes are more vast it is better to override `make_train_dataloader`.

In general, if it sufficient to the user, overriding `_obtain_common_dataloader_parameters` instead of `make_train_dataloader` seems better so that old code can exploit future improvements in `make_train_dataloader`.

File: avalanche/training/plugins/gdumb.py [hunk 11]
```diff
@@ -21,8 +21,14 @@ class GDumbPlugin(SupervisedPlugin):
...
     """
 
-    def __init__(self, mem_size: int = 200):
-        super().__init__()
+    def __init__(
+            self, 
+            mem_size: int = 200,
+            supports_distributed: Optional[bool] = None):
+        super().__init__(
+            supports_distributed=self._check_distributed_support(
```
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/training/plugins/gdumb.py:29]: See the long explanation above

File: avalanche/distributed/distributed_consistency_verification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/distributed/distributed_consistency_verification.py:18]: It seems that the only avalanche-specific hash function in that file is `hash_benchmark`. Do you think it is still appropriate to move it to `CLScenario`?

File: avalanche/core.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/core.py]: Updated in last commit. It now uses this: https://peps.python.org/pep-0487/

File: avalanche/training/templates/base_sgd.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/training/templates/base_sgd.py]: Improved doc with last commit.

File: avalanche/distributed/distributed_consistency_verification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AntonioCarta) [reply avalanche/distributed/distributed_consistency_verification.py:18]: I think it's better to reuse the class `__hash__` method if possible so that child classes can safely override its behavior if needed. Also, `hash_dataset` should work only on `AvalancheDataset` since we don't really support any other dataset.

**Comment by AntonioCarta** [comment]
Ok, I think the `model_device` and `__hash__` are the only remaining issues.

File: avalanche/models/dynamic_modules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lrzpellegrini) [reply avalanche/models/dynamic_modules.py]: Renamed to `_adaptation_device`, changed doc.

File: avalanche/distributed/distributed_consistency_verification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AntonioCarta) [reply avalanche/distributed/distributed_consistency_verification.py:18]: ok, if it's different we can keep it as is. Maybe it will be more clear to me once I see how you use it for distributed training
